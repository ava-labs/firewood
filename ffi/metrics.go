// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
// See the file LICENSE.md for licensing terms.

package ffi

//go:generate go run generate_cgo.go

// #include <stdlib.h>
// #include "firewood.h"
import "C"

import (
	"strings"
	"unsafe"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/expfmt"

	dto "github.com/prometheus/client_model/go"
)

var _ prometheus.Gatherer = (*Gatherer)(nil)

type Gatherer struct{}

func (Gatherer) Gather() ([]*dto.MetricFamily, error) {
	metrics, err := GatherMetrics()
	if err != nil {
		return nil, err
	}

	reader := strings.NewReader(metrics)

	var parser expfmt.TextParser
	parsedMetrics, err := parser.TextToMetricFamilies(reader)
	if err != nil {
		return nil, err
	}

	lst := make([]*dto.MetricFamily, 0, len(parsedMetrics))
	for _, v := range parsedMetrics {
		lst = append(lst, v)
	}

	return lst, nil
}

// Starts global recorder for metrics.
// This function only needs to be called once.
// An error is returned if this method is called a second time, or if it is
// called after StartMetricsWithExporter.
func StartMetrics() error {
	result := C.fwd_start_metrics()
	return errorFromValue(&result)
}

// Start global recorder for metrics along with an HTTP exporter.
// This function only needs to be called once.
// An error is returned if this method is called a second time, if it is
// called after StartMetrics, or if the exporter failed to start.
func StartMetricsWithExporter(metricsPort uint16) error {
	result := C.fwd_start_metrics_with_exporter(C.uint16_t(metricsPort))
	return errorFromValue(&result)
}

// Collect metrics from global recorder
// Returns an error if the global recorder is not initialized.
// This method must be called after StartMetrics or StartMetricsWithExporter
func GatherMetrics() (string, error) {
	result := C.fwd_gather()
	b, err := bytesFromValue(&result)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// LogConfig configures logs for this process.
type LogConfig struct {
	Path        string
	FilterLevel string
}

// Starts global logs.
// This function only needs to be called once.
// An error is returned if this method is called a second time.
func StartLogs(config *LogConfig) error {
	args := &C.struct_LogArgs{}
	if config.Path != "" {
		args.path = C.CString(config.Path)
		defer C.free(unsafe.Pointer(args.path))
	}
	if config.FilterLevel != "" {
		args.filter_level = C.CString(config.FilterLevel)
		defer C.free(unsafe.Pointer(args.filter_level))
	}
	result := C.fwd_start_logs(args)
	return errorFromValue(&result)
}
