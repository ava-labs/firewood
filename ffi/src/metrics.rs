use std::{
    fs::File,
    io::Write,
    ops::Deref,
    path::Path,
    sync::{atomic::Ordering, Arc, Mutex, Once},
    thread::sleep,
};

use chrono::Utc;

use metrics::Key;
use metrics_util::registry::{AtomicStorage, Registry};

static INIT: Once = Once::new();

pub(crate) fn setup_metrics(file: &Path, metric_interval_secs: u32) {
    INIT.call_once(|| {
        let file = File::create(file).expect("failed to create metrics file");
        let inner = TextRecorderInner {
            registry: Registry::atomic(),
            file: file.into(),
        };
        let recorder = TextRecorder {
            inner: Arc::new(inner),
        };
        metrics::set_global_recorder(recorder.clone()).expect("failed to set recorder");
        std::thread::spawn(move || loop {
            sleep(std::time::Duration::from_secs(metric_interval_secs as u64));
            let mut guard = recorder.inner.file.lock().expect("poisoned");
            writeln!(guard, "{}", Utc::now()).unwrap();
            let counters = recorder.registry.get_counter_handles();
            for (key, counter) in counters {
                writeln!(guard, "{} = {}", key, counter.load(Ordering::Relaxed)).unwrap();
            }
            writeln!(guard).unwrap();
        });
    });
}

#[derive(Debug)]
struct TextRecorderInner {
    registry: Registry<Key, AtomicStorage>,
    file: Mutex<File>,
}

#[derive(Debug, Clone)]
struct TextRecorder {
    inner: Arc<TextRecorderInner>,
}

impl Deref for TextRecorder {
    type Target = Arc<TextRecorderInner>;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl metrics::Recorder for TextRecorder {
    fn describe_counter(
        &self,
        _key: metrics::KeyName,
        _unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
    }

    fn describe_gauge(
        &self,
        _key: metrics::KeyName,
        _unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
    }

    fn describe_histogram(
        &self,
        _key: metrics::KeyName,
        _unit: Option<metrics::Unit>,
        _description: metrics::SharedString,
    ) {
    }

    fn register_counter(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Counter {
        self.inner
            .registry
            .get_or_create_counter(key, |c| c.clone().into())
    }

    fn register_gauge(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Gauge {
        self.inner
            .registry
            .get_or_create_gauge(key, |c| c.clone().into())
    }

    fn register_histogram(
        &self,
        key: &metrics::Key,
        _metadata: &metrics::Metadata<'_>,
    ) -> metrics::Histogram {
        self.inner
            .registry
            .get_or_create_histogram(key, |c| c.clone().into())
    }
}
