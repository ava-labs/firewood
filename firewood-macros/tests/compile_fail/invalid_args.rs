// Test that metrics macro fails with invalid arguments
use firewood_macros::metrics;

#[metrics(123)]
fn function_with_invalid_arg() -> Result<(), &'static str> {
    Ok(())
}

fn main() {}
