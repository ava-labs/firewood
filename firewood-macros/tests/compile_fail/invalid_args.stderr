error: Expected string literal(s) for metric name and optional description
 --> tests/compile_fail/invalid_args.rs:4:11
  |
4 | #[metrics(123)]
  |           ^^^
